nomad/state/state_store.go

```diff
@@ -1890,7 +1890,18 @@ func (s *StateStore) nestedUpdateAllocFromClient(txn *memdb.Txn, index uint64, a
 	copyAlloc.ClientStatus = alloc.ClientStatus
 	copyAlloc.ClientDescription = alloc.ClientDescription
 	copyAlloc.TaskStates = alloc.TaskStates
+
+	// Merge the deployment status taking only what the client should set
+	oldDeploymentStatus := copyAlloc.DeploymentStatus
 	copyAlloc.DeploymentStatus = alloc.DeploymentStatus
+	if oldDeploymentStatus != nil {
+		if oldDeploymentStatus.Canary {
+			copyAlloc.DeploymentStatus.Canary = true
+		}
+		if oldDeploymentStatus.Timestamp.After(copyAlloc.DeploymentStatus.Timestamp) {
```

Comment:
Do we need to do a time based comparison here? Seems like time adjustments on the client could break this comparison. While that might be exceedingly rare what would the outcome be?

Reply:
Hm yeah I can remove it. It is actually not necessary as the case where the client sets it and the server sets it are mutually exclusive, so it doesn't matter.